mitiq/collector.py

```diff
@@ -21,20 +21,29 @@
 from typing import Any, Callable, Iterable, List, Sequence, Tuple, Union
 
 import numpy as np
+from cirq import Result as MeasurementResult  # TODO: Generalize frontend.
 
 from mitiq import QPROGRAM
-from mitiq.interface import (
-    convert_from_mitiq,
-    convert_to_mitiq,
-)
+from mitiq.interface import convert_from_mitiq, convert_to_mitiq
+
+# An `executor` function inputs a quantum program and outputs an object from
+# which expectation values can be computed. Explicitly, this object can be one
+# of the following types:
+QuantumResult = Union[
+    float,  # The expectation value itself.
+    MeasurementResult,  # Sampled bitstrings.
+    # TODO: Support the following:
+    # np.ndarray,  # Density matrix.
+    # Sequence[np.ndarray],  # Wavefunctions sampled via quantum trajectories.
+]
```

Comment:
Initially we planned to use a universal Python dictionary to return raw counts, instead of supporting many independent platform dependent objects. 

However, I actually like what you are proposing (it is easier for the user), as long as we are aware that this will cost  more work from our side (conversions).

Reply:
Probably both a `dict[int, int]` and `Counter[int]` will be recognized as `QuantumResult`s eventually. It's very easy to convert these to/from `cirq.Result`, `qiskit.Whatever`, etc., much easier than converting circuits.